src/sage/rings/polynomial/multi_polynomial_ring_base.pyx

```diff
@@ -1794,3 +1831,113 @@ def unpickle_MPolynomialRing_generic(base_ring, n, names, order):
     from .polynomial_ring_constructor import PolynomialRing
 
     return PolynomialRing(base_ring, n, names=names, order=order)
+
+class MatrixPolynomialAction(Action):
+    def __init__(self, MS, PR):
+        """
+        Initialize a MatrixPolynomialAction object.
+
+        Parameters:
+        MS: The matrix space for the action.
+        PR: The polynomial ring where the action takes place.
+
+        Examples:
+        sage: G = groups.matrix.Sp(4,GF(2))
+        sage: R.<w,x,y,z>=GF(2)[]
+        sage: p=x*y^2 + w*x*y*z + 4*w^2*z+2*y*w^2
+        sage: g=G.1
+        sage: from operator import matmul
+        sage: A = p.parent()._get_action_(g.parent(), matmul, False)
+        sage: TestSuite(A).run()
+        """
+        self._poly_vars = PR.gens()
+        self._vars_vector = MatrixConstructor(self._poly_vars).transpose()
+        self.MS = MS
+        self.PR = PR
+        super().__init__(MS, PR, op=matmul)
+
+    def _act_(self, mat, polynomial):
+        """
+        Perform the action of a matrix on a polynomial.
+
+        Parameters:
+        mat: The matrix to act on the polynomial.
+        polynomial: The polynomial to be acted upon.
+
+        Returns:
+        Polynomial: The result of the action.
+
+        Examples:
+        sage: from sage.rings.polynomial.multi_polynomial_ring_base import MatrixPolynomialAction
+        sage: R.<x, y, z> = PolynomialRing(GF(2), 3)
+        sage: M = Matrix(GF(2), [[1, 1, 0], [0, 1, 1], [1, 0, 1]])
+        sage: p = x*y + y*z + z^2
+        sage: A = MatrixPolynomialAction(M.parent(), R)  # using M.parent() to get the matrix space
+        sage: A._act_(M, p)
+        x^2 + y^2
+        """
+        assert mat.base_ring() == polynomial.base_ring()
```

Comment:
It would be better to check the base ring agreement in the `__init__` and fail there. Although I think we only need a coercion from the matrices base ring into the polynomial base ring (e.g., matrices over ZZ have a natural action on polynomials over QQ).

Here, I think we should make sure the matrix is an element of `self.MS` and similarly for the polynomial. (Although that should already be the case by the time this method is called.)

This assert is also too strong as we only look for coercions of base rings, not equality.

Reply:
Hi, 

I understand what you mean about moving the check to __init__, but some of the mathematical formalisms are a bit lost on me regarding the coercion check. I'm assuming I need to call has_coerce_map_from somehow, but I'm not sure what I need to call it on exactly. Any chance you could be a little more specific - my apologies for not understanding this better.

Reply:
No problem; there is a lot going on with this.

Compare with the `_get_action_()`; you will see that you want `polynomial.base_ring().has_coerce_map_from(mat.base_ring())` to be `True`. If you add the doctest I requested, you should see a failure once you do the action (unless I have missed something) before doing this change.

You can remove the `assert` afterwards, although if you really want a safety check (which really should not be needed at all) is `mat.parent() is self.MS and polynomial.parent() is self.PR`. I wouldn't add it though.

Reply:
ok I've added 
```
assert PR.has_coerce_map_from(MS)
```
to the start of `__init__` and now it's throwing errors - assuming I've misunderstood somehow?

Reply:
It should be the base rings:
```python
assert PR.base_ring().has_coerce_map_from(MS.base_ring())
```